src/edx/mod.rs

```diff
@@ -0,0 +1,427 @@
+//! A Dictionary Encryption Scheme securely stores constant length values inside
+//! an Encrypted Dictionary (EDX).
+
+use std::{
+    collections::{HashMap, HashSet},
+    fmt::Debug,
+    hash::Hash,
+};
+
+use cosmian_crypto_core::reexport::rand_core::CryptoRngCore;
+use zeroize::ZeroizeOnDrop;
+
+pub mod chain_table;
+pub mod entry_table;
+mod structs;
+
+pub use structs::{EncryptedValue, Seed};
+
+use crate::{CallbackErrorTrait, Label, TOKEN_LENGTH};
+
+pub trait TokenDump {
+    type Token;
+
+    type Error;
+
+    async fn dump_tokens(&self) -> Result<HashSet<Self::Token>, Self::Error>;
+}
+
+pub trait DxEnc<const VALUE_LENGTH: usize>: Sync + Send {
+    /// Seed used to derive the key.
+    type Seed: Sized + ZeroizeOnDrop + AsRef<[u8]> + Default + AsMut<[u8]> + Send + Sync;
+
+    /// Cryptographically secure key.
+    type Key: Sized + ZeroizeOnDrop + Send + Sync;
+
+    /// Type of error returned by the scheme.
+    type Error: std::error::Error + Sync + Send;
+
+    /// Cryptographically secure token used to index values inside the encrypted
+    /// dictionary.
+    type Token: Copy
+        + Debug
+        + Hash
+        + Eq
+        + Sized
+        + Send
+        + Sync
+        + From<[u8; TOKEN_LENGTH]>
+        + AsRef<[u8]>;
+
+    /// Fixed length encrypted value stored inside the encrypted dictionary.
+    type EncryptedValue: Debug + Sized + Send + Sync + Clone;
+
+    type Store: EdxStore<VALUE_LENGTH>;
+
+    /// Instantiates a new Dx-Enc scheme.
+    fn setup(edx: Self::Store) -> Self;
+
+    /// Generates a new random seed.
+    fn gen_seed(&self, rng: &mut impl CryptoRngCore) -> Self::Seed;
```

Comment:
Is the seed supposed to be cryptographically secure, i.e. generated by a CSPRNG ?

Reply:
It is since it is used to generate the keys. The RNG passed here implements `CryptoRngCore` which makes it a CSPRNG.